### PR TITLE
[RFC] Builtin FSMonitor Daemon

### PR DESCRIPTION
**Pull request by jeffhostetler**
### **WARNING:**
- This PR currently includes the Simple IPC patch series.  This will
be split out and I'll just directly reference the `git:jh/simple-ipc`
branch after it appears on the mailing list. 
- Also, this PR includes a few PR/CI fixes that should not be 
necessary, but were at one point to get various pipelines to 
build; these will be removed before I submit this.
- So please ignore the first 13 commits (everything prior to the
FSMonitor documentation).
### **CUT ABOVE**

This patch series adds a builtin FSMonitor Daemon to Git that monitors
the working directory for changes using platform-specific file system 
notifications. Client commands, such as `git status`, communicate with
the daemon over a well-known named pipe or Unix domain socket within
the .git directory using the "Simple IPC" facility.

Each instance of `git fsmonitor--daemon` is limited to watching a single
working directory.  Each working directory may have its own daemon instance.

This RFC version includes support for Windows and MacOS file system
events.  A Linux version will be submitted in a later patch series.

The FSMonitor Daemon feature contains the following parts:

1. `fsmonitor-ipc.c`:  Routines containing the `fsmonitor_ipc__` prefix are
     used by client processes to send commands and requests to the
     `git fsmonitor--daemon` daemon process.  Clients communicate directly
     with the daemon process using "Simple IPC" and DO NOT use the Hook API.

2. `builtin/fsmonitor--daemon.c`: `git fsmonitor--daemon [<other-modes>]`
     acts as a client and sends commands and/or requests to the daemon
     (if started).

3. `builtin/fsmonitor--daemon.c`: `git fsmonitor--daemon [--run | --start]` 
     creates a daemon process watching the current working directory.
     This explicitly starts a daemon if the working directory is not already
     being watched.  Normally, commands like `git status` will implicitly
     start the daemon if necessary.

3.1. The daemon process creates a "Simple IPC" thread pool using
     `ipc_server_run_async()`.  These threads receive client connections
     route requests to the `fsmonitor--daemon.c:handle_client` callback.

3.2. The daemon process also creates a thread to watch for file system
     events using `fsmonitor_fs_listen__` prefixed routines.

3.3. `compat/fsmonitor/fsmonitor-fs-listen-<platform>.c`: Routines
     containing the `fsmonitor_fs_listen__` prefix are platform-specific and
     implement the low level file system monitoring.  Linkage is controlled
     by the `FSMONITOR_DAEMON_BACKEND` macro in `config.mak.uname`
     and `Makefile`.

3.4. The platform code uses `fsmonitor_publish` to add newly changed
    files to a token-based queue.

3.5. `handle_client` callbacks (running on random threads) respond to
     clients with deduped lists of changed paths.

### CODE REVIEW

**Comment by dscho** [comment]
>  DROPME! fix set-env for FAILED_TEST_ARTIFACTS 

This shouldn't be necessary anymore. Also, maybe we want to split out the simple IPC stuff into its own PR?

File: fsmonitor.c [hunk 1]
```diff
@@ -147,6 +147,7 @@ void write_fsmonitor_extension(struct strbuf *sb, struct index_state *istate)
 static int query_fsmonitor(int version, const char *last_update, struct strbuf *query_result)
 {
 	struct child_process cp = CHILD_PROCESS_INIT;
+	int result;
```
Comment (dscho) [inline fsmonitor.c:150]: If you use `git commit --fixup <commit>`, it will use the syntax that `git rebase --autosquash` understands...

Comment (dscho) [reply fsmonitor.c:150]: (You could also simply replace the colon by an exclamation point in the commit message)

File: compat/fsmonitor/fsmonitor-fs-listen-win32.c [hunk 2]
```diff
@@ -149,20 +147,30 @@ static void destroy_watch(struct one_watch *watch)
...
-		FILE_NOTIFY_CHANGE_LAST_WRITE |
-		FILE_NOTIFY_CHANGE_CREATION,
-		&watch->count, &watch->overlapped, NULL);
+		watch->hDir, watch->buffer, watch->buf_len, TRUE,
+		dwNotifyFilter, &watch->count, &watch->overlapped, NULL);
+
+	if (!watch->is_active &&
+	    GetLastError() == ERROR_INVALID_PARAMETER &&
+	    watch->buf_len > MAX_BUF_WHEN_NETWORKED) {
+		watch->buf_len = MAX_BUF_WHEN_NETWORKED;
```
Comment (dscho) [inline compat/fsmonitor/fsmonitor-fs-listen-win32.c:171]: I have a hunch that a smaller buffer may be required under different circumstances, too. This code would probably catch that, too, but the name `MAX_BUF_WHEN_NETWORKED` might be too misleading? How about `FALLBACK_MAX_BUF`?

File: TODO [hunk 3]
```diff
@@ -0,0 +1,29 @@
+- Set permissions on the named pipe/Unix socket correctly. No need to let
+  random users spy on worktrees' activity to which they have no read access.
```
Comment (dscho) [inline TODO:2]: You seem to have addressed that in a reply in the Simple-IPC thread.

File: TODO [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+- Skip reporting directories (meaning: make sure we know the type of a path)
+  - To do that, be smart and look whether a path refers to a file in a
+    subdirectory, then make sure that the subdirectory is marked as such
+  - Maybe over-report, and only mark known directories?
+  - Or not worry about this for now, it's okay to report directories, use
+    Trace2 to measure the impact of this
```
Comment (dscho) [inline TODO:9]: I guess there is no harm in reporting directories.

File: TODO [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+- Exclude submodules somehow.
```
Comment (dscho) [inline TODO:11]: For the moment, we do not need to worry about this. It only results in over-reporting, so it does not result in _incorrect_ results, but if anybody using submodules cares enough, that might be a valuable thing to do in the future.

File: TODO [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+- Handle when a file is removed and a directory added in its place: this needs
+  to be reported as "deleted".
```
Comment (dscho) [inline TODO:14]: I'm not quite sure how we deal with this. @jeffhostetler do you know off the top of your head?

File: TODO [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+- Handle when a directory is renamed: only the directory name is reported, but
+  we need to report all paths.
```
Comment (dscho) [inline TODO:17]: I could imagine that we're doing the wrong thing here. Do we have tests for that yet?

File: TODO [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+- Possibly prime the daemon's `paths` hashmap by reading the index and adding
+  all tracked files (and untracked files, if the untracked extension is
+  present)
```
Comment (dscho) [inline TODO:21]: I guess your idea was to do a lazy `git status` (that writes an updated index) after a `git checkout` either via `git maintenance` or some other means. That sounds like a pretty cheap and robust solution.

File: TODO [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+- Use the `untracked` cache to initialize the set of paths to report upon
+  startup.
```
Comment (dscho) [inline TODO:24]: I guess we opted against that.

File: TODO [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+- Consider maybe only supporting this fsmonitor when an untracked cache is
+  present.
```
Comment (dscho) [inline TODO:27]: Without using the untracked cache at start-up, this does not make sense.

File: TODO [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+- Document why we're not listening for `.git/index`.
```
Comment (dscho) [inline TODO:29]: Did we do that yet?

**Review by dscho — COMMENTED** [review]
A few comments/questions on my TODOs before we drop that commit.

File: TODO (hunk 5 above)
Comment (jeffhostetler) [reply TODO:11]: I'll add a caveat to the man page for the daemon.

Comment (jeffhostetler) [reply TODO:11]: and an task to create submodule tests.

File: TODO (hunk 6 above)
Comment (jeffhostetler) [reply TODO:14]: @dscho  I'm not sure what you meant in your original TODO item.  Currently we will see the FS delete event for the path and report that path (without a trailing slash) back to the client (without any indication of why or what happened), so the client will have to look it up in the cache-entries.  We should also get an event for the directory creation which the daemon should report to the client (with a trailing slash) and that should trigger stuff in the untracked-cache.  So I think we're good here.  I'll add a task to create unit tests to confirm this.

File: TODO (hunk 7 above)
Comment (jeffhostetler) [reply TODO:17]: I think we get FS events for the old-name and the new-name on all platforms.  The daemon cannot tell the client that a rename happened, but it should say that both pathnames are dirty.  The client should, for example, drop untracked-cache entries for the old path and create new ones for the new path (as opposed to repurposing the existing objects).
So I think we're good, but I'll add a task to create unit tests to confirm this.

File: TODO (hunk 8 above)
Comment (jeffhostetler) [reply TODO:21]: @dscho  By "prime the daemon's paths hashmap" did you mean the `strintern()` map that we're using or something else?   We only use that to cheaply de-dup paths that we send to the client.

Yes, the first few Git commands will be slow until one of them updates the index with a token that demonstrates that we have continuous observations of the file system with the currently active daemon.

File: TODO (hunk 9 above)
Comment (jeffhostetler) [reply TODO:24]: @dscho  I'm not sure what you mean here.  We are currently the set of paths that the daemon observed.  Your comment here sounds like a design idea that we never discussed.  We should talk about this.

File: TODO (hunk 11 above)
Comment (jeffhostetler) [reply TODO:29]: I'm adding a comment in the path classification code to talk about that.

File: builtin/fsmonitor--daemon.c [hunk 12]
```diff
@@ -1438,6 +1439,13 @@ int cmd_fsmonitor__daemon(int argc, const char **argv, const char *prefix)
 #else
 int cmd_fsmonitor__daemon(int argc, const char **argv, const char *prefix)
 {
+	struct option options[] = {
+		OPT_END(),
+	};
+
+	if (argc == 2 && !strcmp(argv[1], "-h"))
+		usage_with_options(builtin_fsmonitor__daemon_usage, options);
+
```
Comment (dscho) [inline builtin/fsmonitor--daemon.c:1448]: This should probably be squashed into the commit that adds `builtin/fsmonitor--daemon.c`, and for readability I would switch it to an `#ifndef HAVE_FSMONITOR_DAEMON_BACKEND` and have the shorter conditional arm come first.

**Review by dscho — COMMENTED** [review]
I guess we need to push a temporary branch to `gitgitgadget/git` that merges `jh/simple-ipc` and `jh/fsmonitor-prework` so that we can re-target the PR.

File: builtin/fsmonitor--daemon.c (hunk 12 above)
Comment (jeffhostetler) [reply builtin/fsmonitor--daemon.c:1448]: Stolee added that commit into the microsoft/git branch when he was preparing the experimental release.  i dropped it on top of my fsmonitor3 branch for completeness.

My fsmonitor4 branch already has this squashed down into the commit that added the surrounding code.

Comment (jeffhostetler) [reply builtin/fsmonitor--daemon.c:1448]: I should close this fsmonitor3 branch on GGG.  I was going to create a new one for fsmonitor4 when i get ready to send part 3 to the mailing list.

Comment (jeffhostetler) [reply builtin/fsmonitor--daemon.c:1448]: I've already done those merges in the fsmonitor4 branch.  (sorry for the confusion.)

Comment (dscho) [reply builtin/fsmonitor--daemon.c:1448]: Of course you did, and I even saw that! My bad, sorry.

Comment (dscho) [reply builtin/fsmonitor--daemon.c:1448]: Hey @jeffhostetler! Now that `jh/simple-ipc` made it into `next`, you could rebase this PR onto `next` (if you hit the `Edit` button next to the PR title, you can change the target branch). What do you think?

Comment (jeffhostetler) [reply builtin/fsmonitor--daemon.c:1448]: um, I should rebase fsmonitor4 onto next and get rid of those merge commits that i added.  i've been holding fsmonitor3 constant to reflect the experimental release that we made.

I do have a few commits to push to simple-ipc for some last minute change suggestions from Friday.  i should be able to do those in parallel.  i'll look at doing that tomorrow.  (my day got away from me today.)

**Comment by jeffhostetler** [comment]
https://github.com/gitgitgadget/git/pull/900

I'm going to close this PR (based on fsmonitor3) and shift to #900.

**Pull request closed without merging**